### ISSUE TITLE
openapi-explorer fails standard accessibility checks

### ISSUE THREAD

**Issue by ahamelers** [Authress-Engineering/openapi-explorer#274]
I was looking at this as a replacement for swagger UI, which has its own accessibility failures. However, this utility is also extremely inaccessible in some very basic (and hopefully easily fixable) ways.

- Rather than using traditional links or buttons, every menu item is a div, with no aria roles, and no tabindexes, and thus no ability to navigate the menus with keyboard controls
- Focus styles for most buttons are nonexistant, so the user cannot tell what is in focus when they _are_ able to use the keyboard to navigate
- Pretty much every text input is missing a label

There are probably other, less obvious accessibility issues, but these are the major ones I can see immediately. Using accessibility tools and checkers on the code created by this tool, and trying to navigate the built screen using keyboards, would go a long way toward discovering these and any additional problems.

Currently, anyone using this tool on a website meant for public access in countries with legal web accessibility requirements (such as the United States) runs the risk of being sued.

**Comment by wparad** [comment]
We'd be happy to receive a list of specific improvements or a comprehensive report we can use to improve the tool. We'd also support multiple PRs to help resolve any concerns you might have.

However, I will say, as this is a developer tool, the overlap with most accessibility requirements is almost fundamentally nil. So hopefully that isn't going to be a concern for most sites.

In any case, if you could add that list to the ticket so that we can prioritize the improvements that would be greatly appreciated.

**Comment by ahamelers** [comment]
> However, I will say, as this is a developer tool, the overlap with most accessibility requirements is almost fundamentally nil. So hopefully that isn't going to be a concern for most sites.

There in fact exist developers with accessibility needs, and under law in the US, any business open to the public (under the ADA), and any organization that is part of federal and most state governments, or who receive federal funding, or contract with the government (under Section 508) are legally required to meet accessibility standards for their content and tools. Under Section 508, employees with disabilities (including developers) of any of the Section 508 types of organizations, as well as members of the public with disabilities, must be provided with comparable access to all the information as those without disabilities. Many, many, APIs—some of the most used APIs in the world, including those from all universities receiving federal funding, health services organizations including the NIH and CDC, other government entities like NASA, weather services, etc. —and all their documentation fall under these categories of use. And those are only in the US—around 40 other countries and the EU have similar legal requirements.

I realize this is an open source project, probably with a small development team, and cannot meet every need. So if you are unable to prioritize accessibility in this tool, you are well within your rights to simply close this issue. But the overlap you mention is much greater than you imagine.

**Comment by wparad** [comment]
Sure, and we can definitely support those needs, but we'll need a list of changes to be made to support those engineers. Can you share a list with us in this ticket so we can prioritize making those changes?

**Comment by ahamelers** [comment]
Okay, I will put in a pull request for the major issues, which are:

- The menus are not keyboard navigable or focusable
    - Related, most buttons and focusable elements do not have focus indicators
    - Also related, primary buttons do not have hover indicators
- Form input elements must have labels
- Pages should contain a level-one heading
- Landmarks (specifically the nav bar) should have an accessible name

Accessibility issue reports can most easily be gotten using a browser extension like axe DevTools ([Firefox](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chromewebstore.google.com/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd))

There are several other accessibility problems, but mostly around contrast colors and that kind of thing. Problems like that depend on style, and since the style is pretty completely customizable with openapi-explorer, I think correcting color contrast and other style-dependant issues are not the project's responsibility.
